browser/ui/webui/settings/brave_extensions_manifest_v2_installer.cc

```diff
@@ -0,0 +1,222 @@
+#include "brave/components/constants/brave_services_key_helper.h"
+#include "brave/components/constants/network_constants.h"
+#include "chrome/browser/extensions/crx_installer.h"
+#include "chrome/browser/extensions/extension_install_prompt.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/crx_file/crx_verifier.h"
+#include "components/update_client/update_query_params.h"
+#include "content/public/browser/web_contents.h"
+#include "extensions/browser/crx_file_info.h"
+#include "extensions/common/extension_urls.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "url/gurl.h"
+
+namespace extensions_mv2 {
+
+namespace {
+
+bool IsKnownMV2Extension(std::string_view id) {
+  return kPreconfiguredManifestV2Extensions.contains(id);
+}
+
+GURL GetUpdaterExtensionDownloadUrl(std::string_view extenion_id) {
+  const std::string params[] = {base::JoinString({"id", extenion_id}, "="),
+                                "installsource=ondemand", "uc"};
+  const GURL url(
+      extension_urls::GetWebstoreUpdateUrl().spec() + "?response=redirect&" +
+      update_client::UpdateQueryParams::Get(
+          update_client::UpdateQueryParams::CRX) +
+      "&x=" + base::EscapeQueryParamValue(base::JoinString(params, "&"), true));
+  CHECK(url.is_valid());
+  return url;
+}
+
+GURL GetCrxDownloadUrl(const base::Value::Dict& update_manifest,
+                       std::string_view extension_id) {
+  const auto* gupdate = update_manifest.FindDict("gupdate");
+  if (!gupdate) {
+    return GURL();
+  }
+  const auto* app_list = gupdate->FindList("app");
+  if (!app_list) {
+    return GURL();
+  }
+
+  for (const auto& appv : *app_list) {
+    if (!appv.is_dict()) {
+      continue;
+    }
+    const auto& app = appv.GetDict();
+
+    const auto* id = app.FindString("appid");
+    if (!id || *id != extension_id) {
+      continue;
+    }
+
+    const auto* update_check = app.FindDict("updatecheck");
+    if (!update_check) {
+      break;
+    }
+    const auto* codebase = update_check->FindString("codebase");
+    if (!codebase) {
+      break;
+    }
+    const GURL url(*codebase);
+    if (!url.is_valid() || !brave::ShouldAddBraveServicesKeyHeader(url)) {
+      break;
+    }
+    return url;
+  }
+
+  return GURL();
+}
+
+}  // namespace
+
+ExtensionManifestV2Installer::ExtensionManifestV2Installer(
+    const std::string& extension_id,
+    content::WebContents* web_contents,
+    extensions::WebstoreInstallWithPrompt::Callback callback)
+    : extension_id_(extension_id),
+      web_contents_(web_contents->GetWeakPtr()),
+      callback_(std::move(callback)) {
+  CHECK(IsKnownMV2Extension(extension_id));
+  auto* profile =
+      Profile::FromBrowserContext(web_contents_->GetBrowserContext());
+  url_loader_factory_ = profile->GetURLLoaderFactory();
```

Comment:
don't use profile as a lookup key, pass in the URLLoaderFactory instead

Reply:
Fixed